hal/shared/cellular_sig_perc_mapping.cpp

```diff
@@ -0,0 +1,199 @@
+    -20,   /* CELLULAR_HAL_SIG_REGION_POOR */
+    -19,   /* CELLULAR_HAL_SIG_REGION_BAD */
+    -17,   /* CELLULAR_HAL_SIG_REGION_FAIR */
+    -14,   /* CELLULAR_HAL_SIG_REGION_GOOD */
+    -12,   /* CELLULAR_HAL_SIG_REGION_GREAT */
+    //-3     /* MAX */
+};
+
+const int utran_strn_map[] = {
+    -121,   /* CELLULAR_HAL_SIG_REGION_POOR */
+    -115,   /* CELLULAR_HAL_SIG_REGION_BAD */
+    -105,   /* CELLULAR_HAL_SIG_REGION_FAIR */
+    -95,    /* CELLULAR_HAL_SIG_REGION_GOOD */
+    -85,    /* CELLULAR_HAL_SIG_REGION_GREAT */
+    //-24     /* MAX */
+};
+
+const int gsm_strn_map[] = {
+    -111,   /* CELLULAR_HAL_SIG_REGION_POOR */
+    -107,   /* CELLULAR_HAL_SIG_REGION_BAD */
+    -103,   /* CELLULAR_HAL_SIG_REGION_FAIR */
+    -97,    /* CELLULAR_HAL_SIG_REGION_GOOD */
+    -89,    /* CELLULAR_HAL_SIG_REGION_GREAT */
+    //-48     /* MAX */
+};
+
+int get_scaled_strn(int strn_value, hal_net_access_tech_t access_tech) {
+    int res = 0;
+    int region = -1;
+
+    // The strength value obtained is a number multiplied by 100. Get the original value.
+    strn_value = strn_value / 100;
+
+
+    int gen_strn_map[MAX_NUM_OF_REGION_THRESHOLDS] = {0};
+
+    // Get the mapping data according to the RAT
+    switch (access_tech) {
+        case NET_ACCESS_TECHNOLOGY_GSM:
+        case NET_ACCESS_TECHNOLOGY_EDGE: {
+            memcpy(gen_strn_map,gsm_strn_map, sizeof(gen_strn_map));
+            break;
+        }
+        case NET_ACCESS_TECHNOLOGY_UTRAN: {
+            memcpy(gen_strn_map,utran_strn_map, sizeof(gen_strn_map));
+            break;
+        }
+        case NET_ACCESS_TECHNOLOGY_LTE: {
+            memcpy(gen_strn_map,cat1_strn_map, sizeof(gen_strn_map));
+            break;
+        }
+        case NET_ACCESS_TECHNOLOGY_LTE_CAT_M1:
+        case NET_ACCESS_TECHNOLOGY_LTE_CAT_NB1: {
+            memcpy(gen_strn_map,catm1_strn_map, sizeof(gen_strn_map));
+            break;
+        }
+        default: {
+            res = std::numeric_limits<int32_t>::min();
+            break;
+        }
+    }
+
+    // If RAT was not selected, return
+    if (res == std::numeric_limits<int32_t>::min()) {
+        return res;
+    }
+
+    // Check which region the value falls into. Convert into percetages, but scale it to [0,65535]
+    if (strn_value < gen_strn_map[CELLULAR_HAL_SIG_REGION_POOR]) {
+        return 0*65535;
+    } else if (strn_value >= gen_strn_map[MAX_NUM_OF_REGION_THRESHOLDS-1]) {
+        return 65535;
+    } else {
+        for (int i=MAX_NUM_OF_REGION_THRESHOLDS-2; i>=CELLULAR_HAL_SIG_REGION_POOR; i--) {
+            if (strn_value >= gen_strn_map[i]) {
+                region = i;
+                break;
+            }
+        }
+        // Covert dBm to percetange based on that region and the value from mapping table
+        int new_range = sig_strn_perc_thres;
+        int old_range = gen_strn_map[region+1] - gen_strn_map[region];
+        res = ((((strn_value - gen_strn_map[region])*new_range*65535)/old_range) +
```

Comment:
Nit: perhaps consider splitting into multiple steps for readability to emphasize that we are using region as a base first and then do a linear interpolation within the region for the remainder:
```
const uint32_t hundredPercent = 65535;
const uint32_t regionPercentage = (hundredPercent * region * 100) / (MAX_REGIONS - 1);
const uint32_t percentageWithinRegion = ...;
```

Reply:
Changed. I also tested that the new logic works. 
Test results here. 
https://docs.google.com/spreadsheets/d/1HVk-Vsdg32YZarLwmYC8KWFz1xUMh_ZZ_nGrN8QUPvk/edit?usp=sharing